src/sets.jl

```diff
@@ -807,6 +807,123 @@ end
 
 dimension(set::Complements) = 2 * set.dimension
 
+"""
+    AllDifferent(dimension::Int)
+
+The set corresponding to an all-different constraint.
+
+All expressions of a vector-valued function are enforced to take distinct
+values in the solution: for all pairs of expressions, their values must
+differ. This constraint is only defined for integer-valued expressions.
+
+This constraint is sometimes called `distinct`.
+"""
+struct AllDifferent <: AbstractVectorSet
+    dimension::Int
+end
+
+"""
+    Domain{T <: Int}(values::Set{T})
+
+The set corresponding to an enumeration of constant values.
+
+The value of a scalar function is enforced to take a value from this set of
+values.
+
+This constraint is sometimes called `in`.
+"""
+struct Domain{T <: Int} <: AbstractScalarSet
+    values::Set{T}
+end
+
+function Base.copy(set::Domain{T <: Int})
+    return Domain(copy(set.values))
+end
+
+"""
+    Membership(values::Set{AbstractScalarFunction})
+
+The set corresponding to an enumeration of values, given by expressions.
+
+The value of a scalar function is enforced to take a value from this set of
+values. These values are given by expressions potentially depending on other
+variables for the problem.
+
+This constraint is sometimes called `in_set`.
+"""
+struct Membership <: AbstractScalarSet
+    values::Set{AbstractScalarFunction}
+end
```

Comment:
Hmm... We need to think about this one. Set's shouldn't have functions as parameters. Consider the `Complements` and `IndicatorSet`s as examples.

Reply:
How about:
```Julia
"""
    Membership(dimension)

The first element of a function of dimension `dimension` must equal at least 
one of the following `dimension - 1` elements of the function.

## Example

    [x, y, z] in Membership(dimension)
    # enforces `x == y` OR `x == z`.
"""
struct Membership <: AbstractVectorSet
    dimension::Int
end
```